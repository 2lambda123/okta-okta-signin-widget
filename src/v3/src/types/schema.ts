/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import {
  IdxMessage,
  IdxTransaction,
  Input,
  NextStep,
  WebauthnVerificationValues,
} from '@okta/okta-auth-js';
import { IdxOption } from '@okta/okta-auth-js/lib/idx/types/idx-js';
import { FunctionComponent } from 'preact';

import { ClickHandler } from './handlers';
import { ListItem, PasswordSettings } from './password';
import { UserInfo } from './userInfo';

export type FormBag = {
  schema: Record<string, unknown>;
  uischema: UISchemaLayout;
  data: Record<string, unknown>;
  // temp schema bag to handle client validation and form submission
  dataSchema: Record<string, DataSchema>;

};

export type AutoCompleteValue = 'username'
| 'current-password'
| 'one-time-code'
| 'new-password'
| 'tel-national'
| 'given-name'
| 'family-name'
| 'email';

export type InputAttributes = { autocomplete?: AutoCompleteValue; };

// flat params
export type ActionParams = {
  [key: string]: string | boolean | number;
};

export interface ActionOptions {
  actionParams?: ActionParams;
  step: string;
  // deprecate this field once auth-js can support `step` for all scenarios
  action?: NextStep['action'];
}

/**
 * WebAuthNEnrollmentPayload
 */
export type WebAuthNEnrollmentPayload = {
  credentials: {
    /**
     * Represents the client data that was passed
     * to CredentialsContainer.create()
     */
    clientData: string;
    /**
     * BtoA String containing authenticator data and an attestation statement
     * for a newly-created key pair.
     */
    attestation: string;
  }
};

/**
 * WebAuthNVerificationPayload
 */
export type WebAuthNVerificationPayload = {
  credentials: WebauthnVerificationValues
};

export type WebAuthNEnrollmentHandler = (transaction: IdxTransaction) =>
Promise<WebAuthNEnrollmentPayload>;

export type WebAuthNAuthenticationHandler = (transaction: IdxTransaction) =>
Promise<WebAuthNVerificationPayload>;

export interface UISchemaElement {
  scope?: string; // TODO: remove
  type: string;
  label?: string;
}

export interface UISchemaLayout {
  type: UISchemaLayoutType;
  elements: (UISchemaElement | UISchemaLayout)[];
  options?: {
    onClick?: ClickHandler;
  }
}

export type PickerSchema = {
  tester: (schema: UISchemaElement) => boolean;
  mapper?: (schema: UISchemaElement) => UISchemaElement;
};
export interface CustomLayout {
  type: UISchemaLayoutType;
  elements: (CustomLayout | UISchemaElement | PickerSchema)[];
}

export enum UISchemaLayoutType {
  HORIZONTAL = 'HorizontalLayout',
  VERTICAL = 'VerticalLayout',
  STEPPER = 'Stepper',
}

export function isUISchemaLayoutType(type: string) {
  return Object.values(UISchemaLayoutType).includes(type as UISchemaLayoutType);
}

export interface FieldElement extends UISchemaElement {
  name: string;
  // TODO: refactor - dump inputMeta at top level of options
  // TODO: only use limited field as i18n field
  options: {
    inputMeta: Input;
    format?: string;
    attributes?: InputAttributes;
    defaultOption?: string | number | boolean;
    type?: string;
    customOptions?: IdxOption[],
    translations?: TranslationInfo[];
    hint?: string;
  };
}

export interface InputTextElement extends FieldElement {
  options: FieldElement['options'] & {
    attributes: InputAttributes;
  };
}
// https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#attr-type
export enum ButtonType {
  SUBMIT = 'submit',
  BUTTON = 'button',
  RESET = 'reset',
}

// TODO: use type instead of format in tester function
export interface ButtonElement extends UISchemaElement {
  type: 'Button',
  options: ActionOptions & {
    type: ButtonType;
    variant?: 'primary' | 'floating' | 'secondary';
    wide?: boolean;
    deviceChallengeUrl?: string;
    includeData?: boolean;
    dataType?: 'cancel' | 'save';
    dataSe?: string;
    stepToRender?: string;
  };
}

export interface AuthenticatorButtonElement {
  type: 'AuthenticatorButton';
  label: string;
  options: Omit<ButtonElement['options'], 'type' | 'step'> & {
    key: string;
    ctaLabel: string;
    description?: string;
  };
}

export interface WebAuthNButtonElement extends UISchemaElement {
  type: 'WebAuthNSubmitButton';
  options: {
    label: string;
    step: string;
    onClick: (() => Promise<WebAuthNEnrollmentPayload>)
    | (() => Promise<WebAuthNVerificationPayload>)
    submitOnLoad?: boolean;
    showLoadingIndicator?: boolean;
  };
}

export interface TitleElement extends UISchemaElement {
  type: 'Title';
  options: {
    content: string;
  };
}

export interface HeadingElement extends UISchemaElement {
  type: 'Heading';
  options: {
    level: 1 | 2 | 3 | 4 | 5 | 6;
    visualLevel: 1 | 2 | 3 | 4 | 5 | 6;
    content: string;
  };
}

export interface DescriptionElement extends UISchemaElement {
  type: 'Description';
  options: {
    content: string;
  };
}

export interface ReminderElement extends UISchemaElement {
  type: 'Reminder';
  options: {
    /**
     * The call to action text in the reminder content area
     */
    ctaText: string;
    /**
     * Override the default timeout before reminder appears
     */
    timeout?: number;
    linkLabel?: string;
    action?: NextStep['action'];
  };
}

export interface ListElement extends UISchemaElement {
  type: 'List';
  options: {
    items: string[];
    type?: 'unordered' | 'ordered' | 'description';
    description?: string;
  };
}

export interface PasswordRequirementsElement extends UISchemaElement {
  type: 'PasswordRequirements',
  options: {
    id: string;
    header: string;
    userInfo: UserInfo;
    settings: PasswordSettings;
    requirements: ListItem[];
    validationDelayMs: number;
  }
}

export interface LinkElement extends UISchemaElement {
  type: 'Link';
  options: ActionOptions & {
    label: string;
    href?: string;
    dataSe?: string;
  };
}

export interface ImageWithTextElement extends UISchemaElement {
  type: 'ImageWithText';
  options: {
    id: string;
    SVGIcon: FunctionComponent;
    textContent: string;
  };
}

export interface QRCodeElement extends UISchemaElement {
  type: 'QRCode';
  options: {
    label: string;
    data: string;
  };
}

export interface SpinnerElement extends UISchemaElement {
  type: 'Spinner';
  options: {
    label: string;
    valueText: string;
  };
}

export interface InfoboxElement extends UISchemaElement {
  options: {
    message: string;
    class: string;
    contentType: string;
    title?: string;
  }
}

export interface SuccessCallback extends UISchemaElement {
  options: {
    data: Record<string, unknown>;
  }
}

export type StepperNavButtonConfigDirection = 'next' | 'prev';

export type StepperNavButtonConfigAttrs = {
  variant?: 'primary' | 'secondary';
  label?: string;
  id?: string;
};

export interface StepperLayout {
  type: UISchemaLayoutType.STEPPER;
  elements: UISchemaLayout[];
  options?: {
    defaultStepIndex: () => number;
  }
}

export interface StepperButtonElement {
  type: 'StepperButton',
  label: string;
  options: Omit<ButtonElement['options'], 'step'>
  & {
    nextStepIndex: number;
  }
}

export interface StepperRadioElement {
  type: 'StepperRadio',
  options: {
    customOptions: Array<IdxOption & { key?: string; }>,
    name: string;
    defaultOption: string | number | boolean;
  }
}

export interface RedirectElement extends UISchemaElement {
  type: 'Redirect',
  options: { url: string; },
}

type ValidateFunction = (data: FormBag['data']) => Partial<IdxMessage>[]
| undefined;

export interface DataSchema {
  validate?: ValidateFunction;
}

export interface TranslationInfo {
  name: string;
  i18nKey: string;
  value: string;
}

export interface PasswordWithConfirmationElement extends UISchemaElement {
  type: 'PasswordWithConfirmation';
  options: {
    input: FieldElement,
  };
}
