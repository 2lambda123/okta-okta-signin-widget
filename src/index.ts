// Entry point for Typescript build
import OktaSignIn from './widget/OktaSignIn';

export default OktaSignIn;
export { OktaSignIn };

export * from './types';
