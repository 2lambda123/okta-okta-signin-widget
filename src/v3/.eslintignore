!.*.json
!.eslintrc.js
!tsconfig.json
assets/
build/
build2/
coverage/
dist/
mocks/
node_modules/
packages/**/*.js
scripts/
size-plugin.json
src/properties/json/
src/static/mockServiceWorker.js
static/
target
test-reports/
test/unit/helpers/xhr/
vendor/
