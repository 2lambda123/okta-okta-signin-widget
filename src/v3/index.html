<html lang="en">

<head>
  <title>SIW G3 Playground</title>
  <meta charset="UTF-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <!-- FIXME /css/okta-sign-in.next.css -->
  <link href="/style.css" type="text/css" rel="stylesheet" />
  <!-- <script src="/js/okta-sign-in.next.min.js"></script> -->
  <script type="module" src="./src/index.ts"></script>
</head>

<body>
  <div id="okta-login-container"></div>
  <script type="module" src="./src/playground.ts"></script>
</body>

</html>
