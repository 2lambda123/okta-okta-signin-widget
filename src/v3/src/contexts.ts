/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { IdxMessage, IdxTransaction, OktaAuth } from '@okta/okta-auth-js';
import { createContext } from 'preact';
import { MutableRef, StateUpdater, useContext } from 'preact/hooks';

import {
  ActionOptions,
  FormBag,
  WidgetProps,
} from './types';

// Widget context
type IWidgetContext = {
  authClient: OktaAuth;
  widgetProps: WidgetProps;
  setMessages: StateUpdater<IdxMessage[]>;
  // // TODO: OKTA-502849 - Update param type
  // (RenderSuccessCallback / RenderErrorCallback) once merged into okta-signin-widget
  onSuccessCallback?: (data: Record<string, unknown>) => void;
  onErrorCallback?: (data: Record<string, unknown>) => void;
  idxTransaction: IdxTransaction | undefined;
  setIdxTransaction: StateUpdater<IdxTransaction | undefined>;
  stepToRender: string | undefined;
  setStepToRender: StateUpdater<string | undefined>;
  data: Record<string, unknown>;
  setData: StateUpdater<Record<string, unknown>>;
  dataSchemaRef: MutableRef<FormBag['dataSchema'] | undefined>;
};

const createWidgetContext = <T extends unknown>() => {
  // Create a context with a generic parameter or undefined
  const genericContext = createContext<T | undefined>(undefined);

  // Check if the value provided to the context is defined or throw an error
  const useWidgetContext = () => {
    const contextIsDefined = useContext(genericContext);
    if (!contextIsDefined) {
      throw new Error('useWidgetContext must be used within a Provider');
    }
    return contextIsDefined;
  };

  return [useWidgetContext, genericContext.Provider] as const;
};

const [useWidgetContext, WidgetContextProvider] = createWidgetContext<IWidgetContext>();

export { useWidgetContext, WidgetContextProvider };

// Form context
type IFormContext = {
  submissionOptionsRef: MutableRef<ActionOptions | undefined>;
};

export const FormContext = createContext<IFormContext>({} as IFormContext);

export const useFormContext = () => useContext(FormContext);

// Stepper context
type IStepperContext = {
  setStepIndex: StateUpdater<number | undefined>;
};

export const StepperContext = createContext<IStepperContext>({
  setStepIndex: () => {},
});

export const useStepperContext = () => useContext(StepperContext);
