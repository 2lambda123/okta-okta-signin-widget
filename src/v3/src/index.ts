/*
 * Copyright (c) 2022-present, Okta, Inc. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant
 * to the Apache License, Version 2.0 (the "License.")
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { config } from '../widgetrc';
import { OktaSignIn } from './OktaSignIn';

// if (process.env.NODE_ENV === 'development' || process.env.NODE_ENV === 'test') {
//   import('./mocks/browser')
//     .then(({ getWorker }) => getWorker())
//     .then((worker) => worker?.start());
// }

if (import.meta.env.MODE === 'development') {
  // if (import.meta.env.VITE_MSW) {
  //   const { worker } = await import('../mocks/browser');
  //   worker.start({ quiet: import.meta.env.VITE_MSW_QUIET });
  // }
  
  // eslint-disable-next-line no-new
  new OktaSignIn(config);
}
