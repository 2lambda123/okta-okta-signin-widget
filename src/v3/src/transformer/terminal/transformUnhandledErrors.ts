/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { AuthApiError } from '@okta/okta-auth-js';

import {
  FormBag,
  InfoboxElement,
  MessageTypeVariant,
  UISchemaElement,
  WidgetProps,
} from '../../types';
import { createForm } from '../utils';

type ErrorTransformer = (widgetProps: WidgetProps, error?: AuthApiError) => FormBag;

const appendSpecialErrorMessages = (
  elements: UISchemaElement[],
  errorDescription: string,
  error?: string,
) => {
  let message;
  if (error === 'invalid_request' && errorDescription === 'The recovery token is invalid') {
    message = 'oie.invalid.recovery.token';
  } else if (error === 'access_denied' && !!errorDescription) {
    message = 'oie.feature.disabled';
  } else {
    message = 'oie.configuration.error';
  }

  elements.push({
    type: 'InfoBox',
    options: {
      message,
      class: MessageTypeVariant.ERROR,
    },
  } as InfoboxElement);
};

export const transformUnhandledErrors: ErrorTransformer = (widgetProps, error) => {
  const formBag: FormBag = createForm();

  if (!error) {
    formBag.uischema.elements.push({
      type: 'InfoBox',
      options: {
        message: 'oform.error.unexpected',
        class: MessageTypeVariant.ERROR,
        contentType: 'string',
      },
    } as InfoboxElement);
    return formBag;
  }

  if (error && error.xhr && !error.errorSummary) {
    const { xhr } = error;
    if (xhr && xhr.responseText?.includes('messages')) {
      const errorResponse = JSON.parse(xhr.responseText);
      const { messages: { value: [message] } } = errorResponse;
      formBag.uischema.elements.push({
        type: 'InfoBox',
        options: {
          message: message.i18n.key,
          contentParams: message.i18n.params,
          class: MessageTypeVariant.ERROR,
          contentType: 'string',
        },
      } as InfoboxElement);

      // TODO: re-visit, handle side effects in hooks
      // If the session expired, this clears session to allow new transaction bootstrap
      if (message.i18n.key === 'idx.session.expired') {
        const { authClient } = widgetProps;
        authClient?.transactionManager.clear();
      }

      return formBag;
    }
  }

  appendSpecialErrorMessages(formBag.uischema.elements, error.errorSummary, error.errorCode);

  return formBag;
};
