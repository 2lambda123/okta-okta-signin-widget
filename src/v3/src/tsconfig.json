{
  "extends": "../tsconfig.base.json",
  "include": [
    "./**/PhoneAuthenticator.test.tsx"
  ]
}
