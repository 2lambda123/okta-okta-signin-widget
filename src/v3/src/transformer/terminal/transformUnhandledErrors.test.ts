/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { AuthApiError } from '@okta/okta-auth-js';
import { InfoboxElement, MessageTypeVariant, WidgetProps } from 'src/types';

import { transformUnhandledErrors } from './transformUnhandledErrors';

describe('Unhandled Error Transformer Tests', () => {
  let apiError: AuthApiError;
  let mockProps: WidgetProps;

  beforeEach(() => {
    mockProps = {};
    apiError = {
      name: '',
      message: '',
      errorSummary: '',
      errorCode: '',
    };
  });

  it('should add Infobox with unexpected error message when error is not provided', () => {
    const formBag = transformUnhandledErrors(mockProps);

    expect(formBag.uischema.elements.length).toBe(1);
    const el = formBag.uischema.elements[0] as InfoboxElement;
    expect(el.type).toBe('InfoBox');
    expect(el.options?.message).toBe('oform.error.unexpected');
    expect(el.options?.class).toBe(MessageTypeVariant.ERROR);
  });

  it('should add info box when response is invalid recovery token error', () => {
    const mockErrorMessage = 'The recovery token is invalid';
    apiError = {
      ...apiError,
      errorCode: 'invalid_request',
      errorSummary: mockErrorMessage,
    };
    const formBag = transformUnhandledErrors(mockProps, apiError);

    expect(formBag.uischema.elements.length).toBe(1);
    expect(formBag.uischema.elements[0].type).toBe('InfoBox');
    expect((formBag.uischema.elements[0] as InfoboxElement).options?.message).toBe('oie.invalid.recovery.token');
    expect((
      formBag.uischema.elements[0] as InfoboxElement
    ).options?.class).toBe(MessageTypeVariant.ERROR);
  });

  it('should add info box when oie is not enabled error', () => {
    const mockErrorMessage = 'Another mocked error message';
    apiError = {
      ...apiError,
      errorCode: 'access_denied',
      errorSummary: mockErrorMessage,
    };
    const formBag = transformUnhandledErrors(mockProps, apiError);

    expect(formBag.uischema.elements.length).toBe(1);
    const el = formBag.uischema.elements[0] as InfoboxElement;
    expect(el.type).toBe('InfoBox');
    expect(el.options?.message).toBe('oie.feature.disabled');
    expect(el.options?.class).toBe(MessageTypeVariant.ERROR);
  });

  it('should add info box when oie configuration error', () => {
    const mockErrorMessage = 'Yet another mocked error message';
    apiError = {
      ...apiError,
      errorCode: 'some_error_key',
      errorSummary: mockErrorMessage,
    };
    const formBag = transformUnhandledErrors(mockProps, apiError);

    expect(formBag.uischema.elements.length).toBe(1);
    const el = formBag.uischema.elements[0] as InfoboxElement;
    expect(el.type).toBe('InfoBox');
    expect(el.options?.message).toBe('oie.configuration.error');
    expect(el.options?.class).toBe(MessageTypeVariant.ERROR);
  });
});
