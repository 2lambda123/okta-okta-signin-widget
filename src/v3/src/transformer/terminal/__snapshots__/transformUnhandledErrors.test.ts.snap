// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unhandled Error Transformer Tests should add Infobox with unexpected error message when error is not provided 1`] = `
Object {
  "data": Object {},
  "dataSchema": Object {
    "fieldsToValidate": Array [],
  },
  "schema": Object {},
  "uischema": Object {
    "elements": Array [
      Object {
        "options": Object {
          "class": "ERROR",
          "contentType": "string",
          "message": "oform.error.unexpected",
        },
        "type": "InfoBox",
      },
    ],
    "type": "VerticalLayout",
  },
}
`;

exports[`Unhandled Error Transformer Tests should add info box when oie configuration error 1`] = `
Object {
  "data": Object {},
  "dataSchema": Object {
    "fieldsToValidate": Array [],
  },
  "schema": Object {},
  "uischema": Object {
    "elements": Array [
      Object {
        "options": Object {
          "class": "ERROR",
          "message": "oie.configuration.error",
        },
        "type": "InfoBox",
      },
    ],
    "type": "VerticalLayout",
  },
}
`;

exports[`Unhandled Error Transformer Tests should add info box when oie is not enabled error 1`] = `
Object {
  "data": Object {},
  "dataSchema": Object {
    "fieldsToValidate": Array [],
  },
  "schema": Object {},
  "uischema": Object {
    "elements": Array [
      Object {
        "options": Object {
          "class": "ERROR",
          "message": "oie.feature.disabled",
        },
        "type": "InfoBox",
      },
    ],
    "type": "VerticalLayout",
  },
}
`;

exports[`Unhandled Error Transformer Tests should add info box when response is invalid recovery token error 1`] = `
Object {
  "data": Object {},
  "dataSchema": Object {
    "fieldsToValidate": Array [],
  },
  "schema": Object {},
  "uischema": Object {
    "elements": Array [
      Object {
        "options": Object {
          "class": "ERROR",
          "message": "oie.invalid.recovery.token",
        },
        "type": "InfoBox",
      },
    ],
    "type": "VerticalLayout",
  },
}
`;
